sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala

```diff
@@ -374,6 +374,7 @@ abstract class JdbcDialect extends Serializable with Logging {
     case dateValue: Date => "'" + dateValue + "'"
     case dateValue: LocalDate => s"'${DateFormatter().format(dateValue)}'"
     case arrayValue: Array[Any] => arrayValue.map(compileValue).mkString(", ")
+    case binaryValue: Array[Byte] => binaryValue.map("%02X".format(_)).mkString("X'", "", "'")
```

Comment:
Is this good for any databases?

Reply:
This is actually a bug existed before.
Array[Byte] is matched as wildcard instead of Array[Any], therefore literal of binary data is eventually formatted as '[B@757d6814'

This fix is trying to format literal of binary data in hex format which is the case at least for MySQL, PostgreSQL, MS SQL Server and Spark SQL.

Reply:
If so, could we split this fix into another PR?

Reply:
Then it will break the test as there some functions takes binary data, e.g. md5, sha1 and sha2

Reply:
How about this, let me add a test to this type to V2JDBCTest.scala, this way we are sure the change is tested on other databases as well.

Reply:
Please split this one into four PRs at least.
1. Fix bug.
2. supports Cast.
3. supports all the mathmatical functions.
4. others.

Reply:
The PR to fix binary literal is created, PTAL: https://github.com/apache/spark/pull/49335

Existing MySQL connector only support two functions with types that works. I'm trying to figure out a way to make it push cast to MySQL to test cast related code to reproduce the bug without introducing new SQL functions support.

Reply:
I mean fix the bug with a dedicated PR.

Reply:
The PR to fix cast types on MySQL is created: PTAL: #49453

Reply:
This PR has been refactored to depend on #49452 and #49453 PTAL

We need to wait for these two prerequisites merged and rebase on the new base before tests can pass.

Reply:
Let's fix that bug first, then promote this one.